### PR TITLE
HDDS-7372. Fixing missing jars in classpath by specifying jar versions

### PR DESCRIPTION
**Pull request by nishitpatira**
### Problem
There are missing jars in the classpath of some of the modules within ozone. This can cause a risk of `ClassNotFoundException` or `ClassDefNotFoundError` amongst others at runtime.

### Investigation
While analyzing mvn dependency tree and ozone classpaths, we found that the issue was because of [Dependency Mediation](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#transitive-dependencies). Out of multiple conflicting versions of the same artifact, maven selects the version closest to the base package. 

### Resolution
The resolution for this was to clearly specify in the pom rules about which version to select in case of a conflict. 
We achieved this by using `dependencyConvergence` rules. 

1. Enforce dependencyConvergence on the package. This forces the build to fail until a particular version for conflicting dependencies is specified. 
2. Recursively run `mvn clean validate` on all the packages. Find the dependencies which are breaking and select the specific version of the dependency which needs to be resolved to. 

**How was a specific version selected?**
For all the conflicts, we decided to chose the latest version of a given artifact. 
For example, in case of com.squareup.okhttp3:okhttp, the conflicting versions were 4.9.0 and 4.9.3. We decided to pick the latest version., ie. 4.9.3.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-7372

## How was this patch tested?

The patch was tested in 2 ways 
1. by running dependencyConvergence rule (via mvn validate) to ensure that there are no more conflicting dependencies.
2. by running ozone validate --classpath <module> to get a success message that all the jars present in the classpath are present in the share/ozone/lib folder. 

In the logs below, we have the **Before** & **After** section (trimmed for readability). Before section shows that there are jars present on the classpath which are missing from the /lib folder. 

After section shows that all the classpaths are successfully validated and there are no missing jars.

### Before : 

`bash-4.2$ ozone --validate classpath ozone-common ; ozone --validate classpath ozone-csi ; ozone --validate classpath ozone-datanode ; ozone --validate classpath ozone-insight ; ozone --validate classpath ozone-manager ; ozone --validate classpath ozone-recon ; ozone --validate classpath ozone-s3gateway ; ozone --validate classpath ozone-tools 

Validating classpath file: /opt/hadoop/share/ozone/classpath/ozone-common.classpath
ERROR: Jar file /opt/hadoop/share/ozone/lib/okhttp-4.9.0.jar is missing
ERROR: Jar file /opt/hadoop/share/ozone/lib/kotlin-stdlib-common-1.6.21.jar is missing
Validation FAILED due to missing jar files!

Validating classpath file: /opt/hadoop/share/ozone/classpath/ozone-csi.classpath
ERROR: Jar file /opt/hadoop/share/ozone/lib/okhttp-4.9.0.jar is missing
ERROR: Jar file /opt/hadoop/share/ozone/lib/kotlin-stdlib-common-1.6.21.jar is missing
Validation FAILED due to missing jar files!

Validating classpath file: /opt/hadoop/share/ozone/classpath/ozone-datanode.classpath
ERROR: Jar file /opt/hadoop/share/ozone/lib/kotlin-stdlib-common-1.6.21.jar is missing
Validation FAILED due to missing jar files!

Validating classpath file: /opt/hadoop/share/ozone/classpath/ozone-insight.classpath
ERROR: Jar file /opt/hadoop/share/ozone/lib/joda-time-2.10.6.jar is missing
Validation FAILED due to missing jar files!

Validating classpath file: /opt/hadoop/share/ozone/classpath/ozone-manager.classpath
ERROR: Jar file /opt/hadoop/share/ozone/lib/joda-time-2.10.6.jar is missing
Validation FAILED due to missing jar files!

Validating classpath file: /opt/hadoop/share/ozone/classpath/ozone-recon.classpath
ERROR: Jar file /opt/hadoop/share/ozone/lib/joda-time-2.10.6.jar is missing
Validation FAILED due to missing jar files!

Validating classpath file: /opt/hadoop/share/ozone/classpath/ozone-s3gateway.classpath
Validation SUCCESSFUL, all required jars are present!

Validating classpath file: /opt/hadoop/share/ozone/classpath/ozone-tools.classpath
Validation SUCCESSFUL, all required jars are present!
`
### After

`bash-4.2$ ozone --validate classpath ozone-common ; ozone --validate classpath ozone-csi ; ozone --validate classpath ozone-datanode ; ozone --validate classpath ozone-insight ; ozone --validate classpath ozone-manager ; ozone --validate classpath ozone-recon ; ozone --validate classpath ozone-s3gateway ; ozone --validate classpath ozone-tools

Validating classpath file: /opt/hadoop/share/ozone/classpath/ozone-common.classpath
Validation SUCCESSFUL, all required jars are present!

Validating classpath file: /opt/hadoop/share/ozone/classpath/ozone-csi.classpath
Validation SUCCESSFUL, all required jars are present!

Validating classpath file: /opt/hadoop/share/ozone/classpath/ozone-datanode.classpath
Validation SUCCESSFUL, all required jars are present!

Validating classpath file: /opt/hadoop/share/ozone/classpath/ozone-insight.classpath
Validation SUCCESSFUL, all required jars are present!

Validating classpath file: /opt/hadoop/share/ozone/classpath/ozone-manager.classpath
Validation SUCCESSFUL, all required jars are present!

Validating classpath file: /opt/hadoop/share/ozone/classpath/ozone-recon.classpath
Validation SUCCESSFUL, all required jars are present!

Validating classpath file: /opt/hadoop/share/ozone/classpath/ozone-s3gateway.classpath
Validation SUCCESSFUL, all required jars are present!

Validating classpath file: /opt/hadoop/share/ozone/classpath/ozone-tools.classpath
Validation SUCCESSFUL, all required jars are present!
`

(Please explain how this patch was tested. Ex: unit tests, manual tests)
(If this patch involves UI changes, please attach a screen-shot; otherwise, remove this)

### CODE REVIEW

File: hadoop-ozone/ozonefs-hadoop2/pom.xml [hunk 1]
```diff
@@ -53,19 +53,19 @@
...
     </dependency>
     <dependency>
       <groupId>org.apache.hadoop</groupId>
       <artifactId>hadoop-auth</artifactId>
       <scope>provided</scope>
-      <version>2.7.3</version>
+      <version>3.3.4</version>
     </dependency>
     <dependency>
       <groupId>ch.qos.reload4j</groupId>
```
Comment (adoroszlai) [inline hadoop-ozone/ozonefs-hadoop2/pom.xml:71]: This submodule is specific to Hadoop 2.  Please do not update it to Hadoop 3.  If necessary, please add an exception or override for the enforcer plugin.

File: hadoop-ozone/dist/src/main/license/jar-report.txt [hunk 2]
```diff
@@ -1,10 +1,11 @@
+share/ozone/lib/FastInfoset.jar
```
Comment (adoroszlai) [inline hadoop-ozone/dist/src/main/license/jar-report.txt:1]: Please try to avoid reordering the jar-report.  These unnecessary changes make it more difficult to see what has actually changed.

File: hadoop-ozone/pom.xml [hunk 3]
```diff
@@ -257,6 +257,152 @@
...
+      </dependency>
+      <dependency>
+        <groupId>log4j</groupId>
+        <artifactId>log4j</artifactId>
+        <version>1.2.17</version>
+      </dependency>
+      <dependency>
+        <groupId>org.slf4j</groupId>
+        <artifactId>slf4j-log4j12</artifactId>
+        <version>1.7.10</version>
```
Comment (adoroszlai) [inline hadoop-ozone/pom.xml:383]: This seems pretty outdated, we are already using 1.7.36.

File: hadoop-ozone/pom.xml [hunk 4]
```diff
@@ -257,6 +257,152 @@
...
+      <dependency>
+        <groupId>org.apache.curator</groupId>
+        <artifactId>curator-client</artifactId>
+        <version>4.2.0</version>
+      </dependency>
+      <dependency>
+        <groupId>org.xerial.snappy</groupId>
+        <artifactId>snappy-java</artifactId>
+        <version>1.1.8.2</version>
+      </dependency>
```
Comment (adoroszlai) [inline hadoop-ozone/pom.xml:404]: Would it be possible to add these in root POM instead?  Some of these seem to be a duplication of those versions.

**Review by adoroszlai — CHANGES_REQUESTED** [review]
Thanks @nishitpatira for working on this.  In addition to the comments below, I'm still waiting for [CI results in your fork](https://github.com/nishitpatira/ozone/actions/runs/3707352939).

File: hadoop-ozone/ozonefs-hadoop2/pom.xml (hunk 1 above)
Comment (nishitpatira) [reply hadoop-ozone/ozonefs-hadoop2/pom.xml:71]: Reverted back to 2.7.3 and added exclusions for hadoop-auth and hadoop-annotations

File: hadoop-ozone/pom.xml (hunk 3 above)
Comment (nishitpatira) [reply hadoop-ozone/pom.xml:383]: Thanks. 1.7.10 is several years old and I don't know why maven suggested that I pick that up.

**Comment by nishitpatira** [comment]
[Latest CI on personal fork](https://github.com/nishitpatira/ozone/actions/runs/3708782528)

File: hadoop-ozone/pom.xml (hunk 4 above)
Comment (kerneltime) [reply hadoop-ozone/pom.xml]: @adoroszlai I asked @nishitpatira to go per module first to scope the changes, as the last step we can move all the dependencies to the top-level pom and consolidate the versions then.

File: hadoop-ozone/ozonefs-hadoop2/pom.xml [hunk 5]
```diff
@@ -54,6 +54,16 @@
       <artifactId>hadoop-common</artifactId>
       <scope>provided</scope>
       <version>2.7.3</version>
+      <exclusions>
+        <exclusion>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>hadoop-auth</artifactId>
```
Comment (kerneltime) [inline hadoop-ozone/ozonefs-hadoop2/pom.xml:60]: The version for `hadoop-auth` and `hadoop annotations` might be different here as this package is explicitly built to take `dependency` on `hadoop 2.7.3` vs `hadoop3` based packages. We might have to suppress any errors that might arise due to this package depending on a different version of hadoop.

File: pom.xml [hunk 6]
```diff
@@ -72,7 +72,7 @@ xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xs
     <declared.ozone.version>${ozone.version}</declared.ozone.version>
 
     <!-- Apache Ratis version -->
-    <ratis.version>2.4.2-8b8bdda-SNAPSHOT</ratis.version>
+    <ratis.version>2.4.2-8b8bdda-20221130.230506-1</ratis.version>
```
Comment (kerneltime) [inline pom.xml:75]: Where did we get this version string from?

File: hadoop-ozone/pom.xml (hunk 4 above)
Comment (kerneltime) [reply hadoop-ozone/pom.xml]: @nishitpatira if possible can we move all the version definitions to the top-level pom?

Comment (adoroszlai) [reply hadoop-ozone/pom.xml]: This reminds me that we should commit HDDS-7371 first.

**Comment by kerneltime** [comment]
Do take a look at https://github.com/apache/ozone/pull/4059

File: hadoop-ozone/pom.xml (hunk 4 above)
Comment (kerneltime) [reply hadoop-ozone/pom.xml]: Agreed!

File: hadoop-ozone/dist/pom.xml [hunk 7]
```diff
@@ -105,6 +105,16 @@
...
+                <artifactItem>
+                  <groupId>org.apache.hadoop</groupId>
+                  <artifactId>hadoop-auth</artifactId>
+                  <version>2.7.3</version>
+                </artifactItem>
+                <artifactItem>
+                  <groupId>org.apache.hadoop</groupId>
+                  <artifactId>hadoop-annotations</artifactId>
+                  <version>2.7.3</version>
+                </artifactItem>
```
Comment (nishitpatira) [inline hadoop-ozone/dist/pom.xml:117]: To explicitly provide these jars to hadoop-2, explicitly copying them.

File: hadoop-ozone/ozonefs-hadoop2/pom.xml (hunk 5 above)
Comment (nishitpatira) [reply hadoop-ozone/ozonefs-hadoop2/pom.xml:60]: To explicitly provide these jars to hadoop-2, explicitly copying them in `hadoop-dist/pom.xml`

File: pom.xml (hunk 6 above)
Comment (nishitpatira) [reply pom.xml:75]: There were transitive dependencies on `2.4.2-8b8bdda-SNAPSHOT` and `2.4.2-8b8bdda-20221130.230506-1`. There was no help from [mavenCentral](https://central.sonatype.dev/artifact/org.apache.ratis/ratis-tools/2.4.1/overview), so looking at the date, I made a call to use the latter version.

File: hadoop-ozone/dist/src/main/license/jar-report.txt (hunk 2 above)
Comment (nishitpatira) [reply hadoop-ozone/dist/src/main/license/jar-report.txt:1]: I just ran the script and it automatically updated this. I can remove the file, if it's too cumbersome. But, as long as the CI passes, we know that the data in `jar-report.txt` is accurate, no?

File: hadoop-ozone/pom.xml (hunk 4 above)
Comment (nishitpatira) [reply hadoop-ozone/pom.xml]: Moved everything to root pom.

File: hadoop-ozone/dist/src/main/license/jar-report.txt (hunk 2 above)
Comment (adoroszlai) [reply hadoop-ozone/dist/src/main/license/jar-report.txt:1]: You don't need to run `hadoop-ozone/dist/src/main/license/update-jar-report.sh`.  Running `hadoop-ozone/dev-support/checks/dependency.sh` shows you the differences.  Just update `jar-report.txt` manually, keeping current order, inserting/removing items if needed.

File: pom.xml (hunk 6 above)
Comment (kerneltime) [reply pom.xml:75]: Understood. @nishitpatira use `2.4.2-8b8bdda-SNAPSHOT` that are for the same commit hash. The hex code post `2.4.2` is the commit hash. So both are pointing to the same version of the code.

File: hadoop-ozone/ozonefs-hadoop2/pom.xml (hunk 5 above)
Comment (kerneltime) [reply hadoop-ozone/ozonefs-hadoop2/pom.xml:60]: I am not sure I follow why `hadoop-ozone/dist/pom.xml` and this file need to change. When building ozonefs for Hadoop2 we explicitly want to take hadoop2 as a dependency for everything.
Also, do you plan to add the enforcement for `DependencyConvergence`

Comment (kerneltime) [reply hadoop-ozone/ozonefs-hadoop2/pom.xml:60]: I see the problem, we will need to figure out how to keep all hadoop dependencies in `ozonefs-hadoop2` to be with version 2 and the rest of the packages use hadoop3 as their dependency.
@adoroszlai any recommendation?

Comment (nishitpatira) [reply hadoop-ozone/ozonefs-hadoop2/pom.xml:60]: `hadoop-auth` and `hadoop-common` are present in ozonfs-hadoop2 classpath with the version 2.7.3. This was verified by looking at the classpath generated by ```mvn dependency:build-classpath ```. And this jar is being explicitly made available by the copy command in `dist/pom.xml`. For all other modules, we are defaulting to the 3.3.4 version. 
This is exactly what we intended to. Why do you think we still have a problem?

**Comment by kerneltime** [comment]
@smengcl @nandakumar131 @adoroszlai, please take a look as well.

File: hadoop-ozone/dist/pom.xml [hunk 8]
```diff
@@ -105,6 +105,16 @@
                   <artifactId>protobuf-java</artifactId>
                   <version>${grpc.protobuf-compile.version}</version>
                 </artifactItem>
+                <artifactItem>
+                  <groupId>org.apache.hadoop</groupId>
+                  <artifactId>hadoop-auth</artifactId>
+                  <version>2.7.3</version>
```
Comment (smengcl) [inline hadoop-ozone/dist/pom.xml:111]: Try to use `hadoop2.version` instead

```suggestion
                  <version>${hadoop2.version}</version>
```

File: hadoop-ozone/dist/pom.xml [hunk 9]
```diff
@@ -105,6 +105,16 @@
...
+                </artifactItem>
+                <artifactItem>
+                  <groupId>org.apache.hadoop</groupId>
+                  <artifactId>hadoop-annotations</artifactId>
+                  <version>2.7.3</version>
```
Comment (smengcl) [inline hadoop-ozone/dist/pom.xml:116]: Same. Try to use `hadoop2.version` instead?

```suggestion
                  <version>${hadoop2.version}</version>
```

File: hadoop-ozone/dist/pom.xml (hunk 8 above)
Comment (kerneltime) [reply hadoop-ozone/dist/pom.xml:111]: @nishitpatira I can merge the change once you address @smengcl's comments

Comment (nishitpatira) [reply hadoop-ozone/dist/pom.xml]: Made the changes. 
Running CI in https://github.com/nishitpatira/ozone/actions/runs/3753493660

**Review by smengcl — APPROVED** [review]

File: pom.xml [hunk 10]
```diff
@@ -64,6 +64,7 @@ xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xs
   <properties>
     <hadoop2.version>2.7.3</hadoop2.version>
     <hadoop.version>3.3.4</hadoop.version>
+    <hadoop2.version>2.7.3</hadoop2.version>
```
Comment (kerneltime) [inline pom.xml:67]: This is a duplicate of line 65

File: pom.xml [hunk 11]
```diff
@@ -1437,6 +1467,106 @@ xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xs
...
+      <dependency>
+        <groupId>com.nimbusds</groupId>
+        <artifactId>nimbus-jose-jwt</artifactId>
+        <version>9.8.1</version>
+      </dependency>
+      <dependency>
+        <groupId>org.checkerframework</groupId>
+        <artifactId>checker-qual</artifactId>
+        <version>3.12.0</version>
+      </dependency>
```
Comment (adoroszlai) [inline pom.xml:1494]: Please only add version numbers using properties.  (Also applies to few other hard-coded version numbers below.)

File: pom.xml [hunk 12]
```diff
@@ -1437,6 +1467,106 @@ xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xs
...
+      <dependency>
+        <groupId>commons-lang</groupId>
+        <artifactId>commons-lang</artifactId>
+        <version>2.6</version>
+      </dependency>
+      <dependency>
+        <groupId>log4j</groupId>
+        <artifactId>log4j</artifactId>
+        <version>1.2.17</version>
+      </dependency>
```
Comment (adoroszlai) [inline pom.xml:1549]: We have moved away from log4j to reload4j.  Please do not re-introduce this dependency.

**Review by adoroszlai — CHANGES_REQUESTED** [review]

**Comment by kerneltime** [comment]
@nishitpatira can you take a look at the CI failures?

File: pom.xml [hunk 13]
```diff
@@ -1505,22 +1515,22 @@ xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xs
...
         <groupId>org.javassist</groupId>
         <artifactId>javassist</artifactId>
-        <version>3.21.0-GA</version>
+        <version>${javassist.version}</version>
       </dependency>
       <dependency>
         <groupId>org.glassfish.hk2.external</groupId>
         <artifactId>jakarta.inject</artifactId>
-        <version>2.6.1</version>
+        <version>${jakarta.inject.version</version>
```
Comment (tanvipenumudy) [inline pom.xml:1523]: Thank you for the contribution, @nishitpatira! Could you please enclose `${jakarta.inject.version` with a closing curly brace? Thanks.

Comment (nishitpatira) [reply pom.xml:1523]: Hey, sorry, my license for IntelliJ had expired and my local build was failing, so, thought of running it on Actions for getting a better report. Fixed it.

File: pom.xml [hunk 14]
```diff
@@ -1202,6 +1232,11 @@ xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xs
         <artifactId>jul-to-slf4j</artifactId>
         <version>${slf4j.version}</version>
       </dependency>
+      <dependency>
+        <groupId>org.slf4j</groupId>
+        <artifactId>slf4j-log4j12</artifactId>
+        <version>${slf4j.version}</version>
+      </dependency>
```
Comment (adoroszlai) [inline pom.xml:1239]: `slf4j-log4j12` is never used (nor should it be), so this dependency is unnecessary.

(Sorry I had not noticed this in the previous round.)

Comment (nishitpatira) [reply pom.xml:1239]: I added only those dependencies which maven suggested as required.

However, as an experiment, I removed `slf4j-log4j12` but it failed in DependencyConvergence. 

Below is the error 

```
[WARNING] 
Dependency convergence error for org.slf4j:slf4j-log4j12:1.7.10 paths to dependency are:
+-org.apache.ozone:ozone-filesystem-hadoop2:1.3.0-SNAPSHOT
  +-org.apache.hadoop:hadoop-common:2.7.3
    +-org.slf4j:slf4j-log4j12:1.7.10
and
+-org.apache.ozone:ozone-filesystem-hadoop2:1.3.0-SNAPSHOT
  +-org.apache.hadoop:hadoop-common:2.7.3
    +-org.apache.zookeeper:zookeeper:3.5.6
      +-org.slf4j:slf4j-log4j12:1.7.25
and
+-org.apache.ozone:ozone-filesystem-hadoop2:1.3.0-SNAPSHOT
  +-org.apache.hadoop:hadoop-auth:2.7.3
    +-org.slf4j:slf4j-log4j12:1.7.10

[WARNING] Rule 0: org.apache.maven.plugins.enforcer.DependencyConvergence failed with message:
```

File: hadoop-ozone/dist/src/main/license/jar-report.txt (hunk 2 above)
Comment (adoroszlai) [reply hadoop-ozone/dist/src/main/license/jar-report.txt:1]: There still are unnecessarily reordered lines.  On closer look, no changes should be done in the jar report.

File: hadoop-ozone/dist/src/main/license/jar-report.txt [hunk 15]
```diff
@@ -37,23 +38,24 @@ share/ozone/lib/disruptor.jar
...
 share/ozone/lib/guice-assistedinject.jar
 share/ozone/lib/guice-bridge.jar
-share/ozone/lib/guice.jar
 share/ozone/lib/guice-multibindings.jar
 share/ozone/lib/guice-servlet.jar
+share/ozone/lib/guice.jar
 share/ozone/lib/hadoop-annotations.jar
+share/ozone/lib/hadoop-annotations.jar
+share/ozone/lib/hadoop-auth.jar
 share/ozone/lib/hadoop-auth.jar
```
Comment (adoroszlai) [inline hadoop-ozone/dist/src/main/license/jar-report.txt:59]: Duplicate entries are present due to different jar versions for Hadoop 2 and 3.  We should not be including jars from Hadoop 2 in the Ozone distribution.

File: hadoop-ozone/dist/pom.xml [hunk 16]
```diff
@@ -105,6 +105,16 @@
...
+                <artifactItem>
+                  <groupId>org.apache.hadoop</groupId>
+                  <artifactId>hadoop-auth</artifactId>
+                  <version>${hadoop2.version}</version>
+                </artifactItem>
+                <artifactItem>
+                  <groupId>org.apache.hadoop</groupId>
+                  <artifactId>hadoop-annotations</artifactId>
+                  <version>${hadoop2.version}</version>
+                </artifactItem>
```
Comment (adoroszlai) [inline hadoop-ozone/dist/pom.xml:117]: We should not copy Hadoop2 jars to `share/ozone/lib`.  These are expected to be provided by users who wish to use Ozone in a Hadoop2 environment.

**Review by adoroszlai — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 17]
```diff
@@ -1437,6 +1472,101 @@ xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xs
...
+      <dependency>
+        <groupId>net.java.dev.jna</groupId>
+        <artifactId>jna-platform</artifactId>
+        <version>${java.dev.jna.version}</version>
+      </dependency>
+      <dependency>
+        <groupId>org.javassist</groupId>
+        <artifactId>javassist</artifactId>
+        <version>${javassist.version}</version>
+      </dependency>
```
Comment (adoroszlai) [inline pom.xml:1514]: Just noticed this duplicate declaration.

```suggestion
```

https://github.com/apache/ozone/actions/runs/4014300586/jobs/6901462838#step:5:14

**Review by adoroszlai — CHANGES_REQUESTED** [review]
Thanks @nishitpatira for updating the patch.

**Review by adoroszlai — APPROVED** [review]
